rs/xnet/payload_builder/src/certified_slice_pool.rs

```diff
@@ -454,11 +454,14 @@ impl Payload {
         let cutoff = cutoff.unwrap_or_else(|| {
             // `count_bytes()` returned a value above `byte_limit`, but
             // `byte_size` (computed the same way) is below `byte_limit`.
-            panic!(
+            debug_assert!(
+                false,
                 "Invalid `messages_count_bytes`: was {}, expecting {}",
                 messages.count_bytes(),
                 byte_size
-            )
+            );
+            // Cut off before the last message.
+            messages.iter().next_back().unwrap().0.clone()
```

Comment:
Did you consider returning an empty payload instead of making up a cutoff which potentially violates the limits and/or what are the tradeoffs?

Reply:
First of all, this is an error case that could only be triggered by a bug (and would be caught by the `debug_assert` before).

But even so, what would need to happen to reach this point is:
 * the full payload exceeds the byte limit
 * but the incrementally computed payload size is below the byte limit

i.e. we're a bit off in our calculation (or else the whole calculation is so far off that we can't trust it anyway). To put it differently, if we would treat this exact payload as valid when obtained as the prefix of a longer payload, we may as well treat it as valid now.

Reply:
SGTM. Thanks.

My initial concern was about stalling if there is a byte estimate that is somehow off, leading to different estimates in payload building and validation. But I just remembered that we do [this](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/xnet/payload_builder/src/lib.rs?L887) check here in payload building. Meaning that even if we have a bug in the slice pool and end up returning a slice where the byte limit is slightly off, we would either discard the slice in the linked line, or otherwise it is ensured that the byte estimate used in payload validation will be the same as used in payload building.